File: backend/pkg/security/aggregation/max_score.py
from __future__ import annotations

from pkg.apis.security.models import AggregatedDetectionResult, DetectionResult


class MaxScoreAggregator:
    """Research default: a long document is as risky as its highest-scoring chunk."""

    name = "max-score"

    def aggregate(self, results: list[DetectionResult]) -> AggregatedDetectionResult:
        if not results:
            return AggregatedDetectionResult(
                score=0.0,
                is_injection=False,
                chunks_scanned=0,
                detector_results=[],
                metadata={"strategy": self.name},
            )
        top = max(results, key=lambda item: item.score)
        return AggregatedDetectionResult(
            score=top.score,
            is_injection=any(item.is_injection for item in results),
            chunks_scanned=len(results),
            detector_results=results,
            metadata={"strategy": self.name, "top_detector": top.detector},
        )

File: backend/pkg/security/aggregation/maxscore.go
package aggregation

import "github.com/spider/spider/pkg/apis"

type Aggregator interface {
	Name() string
	Aggregate(results []apis.DetectionResult) apis.AggregatedDetectionResult
}

type MaxScoreAggregator struct{}

func (a *MaxScoreAggregator) Name() string { return "max-score" }

func (a *MaxScoreAggregator) Aggregate(results []apis.DetectionResult) apis.AggregatedDetectionResult {
	if len(results) == 0 {
		return apis.AggregatedDetectionResult{
			Score:         0,
			IsInjection:   false,
			ChunksScanned: 0,
		}
	}
	maxScore := 0.0
	isInjection := false
	for _, r := range results {
		if r.Score > maxScore {
			maxScore = r.Score
		}
		if r.IsInjection {
			isInjection = true
		}
	}
	return apis.AggregatedDetectionResult{
		Score:           maxScore,
		IsInjection:     isInjection,
		ChunksScanned:   len(results),
		DetectorResults: results,
	}
}
